### PR TITLE
feat(report): GitHub Dependency Snapshots support

### PR DESCRIPTION
**Pull request by AndreyLevchenko**

### CODE REVIEW

File: pkg/report/genericsbom.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		return xerrors.Errorf("failed to marshal generic sbom: %w", err)
+	}
+
+	if _, err = fmt.Fprint(gsbmw.Output, string(output)); err != nil {
+		return xerrors.Errorf("failed to write generic sbom: %w", err)
+	}
+	return nil
+}
+func buildPurl(pkgName, version, pkgType string) string {
+	return fmt.Sprintf("pkg:/%s/%s@%s", toPurlType(pkgType), url.QueryEscape(pkgName), version)
```
Comment (knqyf263) [inline pkg/report/genericsbom.go:92]: You can use this library.
https://github.com/package-url/packageurl-go

File: pkg/report/genericsbom.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+var resolvedFromVulns = func(result Result) map[string]GsbomPackage {
```
Comment (knqyf263) [inline pkg/report/genericsbom.go:95]: I think we can simply define it as a named function.

File: pkg/report/genericsbom.go (hunk 1 above)
Comment (AndreyLevchenko) [reply pkg/report/genericsbom.go]: fixed

File: pkg/report/genericsbom.go (hunk 2 above)
Comment (AndreyLevchenko) [reply pkg/report/genericsbom.go:97]: fixed

File: pkg/report/genericsbom.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+	if _, err = fmt.Fprint(gsbmw.Output, string(output)); err != nil {
+		return xerrors.Errorf("failed to write generic sbom: %w", err)
+	}
+	return nil
+}
+func buildPurl(pkgName, version, pkgType string) string {
+	purlPackageType := toPurlType(pkgType)
+	var purlNs string
+	//Maven groupid is purl namespace
```
Comment (masahiro331) [inline pkg/report/genericsbom.go:95]: https://github.com/aquasecurity/trivy/pull/1522/files#diff-9bf6e24337d9bac10703d1bf42439dc126b97de4872e8b4a7460fad65b88ff23R95

File: pkg/report/genericsbom.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	if err != nil {
+		return xerrors.Errorf("failed to marshal generic sbom: %w", err)
+	}
+
+	if _, err = fmt.Fprint(gsbmw.Output, string(output)); err != nil {
+		return xerrors.Errorf("failed to write generic sbom: %w", err)
+	}
+	return nil
+}
+func buildPurl(pkgName, version, pkgType string) string {
```
Comment (masahiro331) [inline pkg/report/genericsbom.go:92]: https://github.com/package-url/purl-spec#purl

The GitHub user or organization should be set as the namespace.
Therefore, @ and other characters present in the test case will be escaped.

Comment (AndreyLevchenko) [reply pkg/report/genericsbom.go:92]: I read the link but it's not clear how to extract github user/organization from generic package. I mean there are many types of packages for some there is no obvious way and for some it's simple

File: pkg/report/genericsbom.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		if result.Class == ClassLangPkg {
+			manifest.File = &GsbomFile{
+				SrcLocation: result.Target,
+			}
+		}
+		var resolver depsRslvr
+		if result.Packages != nil {
+			resolver = resolvedFromPackages
+		} else {
+			resolver = resolvedFromVulns
```
Comment (masahiro331) [inline pkg/report/genericsbom.go:74]: If packages are nil, there is no need to create an SBOM.
In order to generate SBOM, it is better to make options such as --list-all-pkgs mandatory or issue a warning.

File: pkg/report/genericsbom.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	resolved := make(map[string]GsbomPackage)
+	for _, pkg := range result.Packages {
+		gsbompkg := GsbomPackage{}
+		gsbompkg.Purl = buildPurl(pkg.Name, pkg.Version, result.Type)
+		resolved[pkg.Name] = gsbompkg
+	}
+	return resolved
+}
+
+func toPurlType(ecosystem string) string {
```
Comment (masahiro331) [inline pkg/report/genericsbom.go:123]: Does Generic SBOM need to support OS packages?
Is this only for language packages?

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**